File: ruoyi-ui/src/views/services/movieService.js
import axios from 'axios';

// 【推荐】将API基础路径定义为常量，方便管理
const API_BASE_URL = 'http://localhost:8080/api/movies';

export const movieService = {
  /**
   * 从后端获取所有电影列表
   * @returns {Promise<Array>} 返回电影数据数组
   */
  async getList() {
    try {
      const response = await axios.get(`${API_BASE_URL}/all`);
      return response.data;
    } catch (error) {
      console.error("Error fetching movie list:", error);
      // 将错误继续向上抛出，让调用它的组件可以处理
      throw error;
    }
  },

  /**
   * 【修正】根据ID从后端获取单个电影的详细信息
   * @param {string|number} id - 电影的ID
   * @returns {Promise<object>} 返回单个电影的数据对象
   */
  async getMovieById(id) {
    // 不再使用 mockMovieDatabase
    try {
      // 构造请求单个电影的URL，例如: http://localhost:8080/api/movies/11
      const response = await axios.get(`${API_BASE_URL}/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Error fetching movie with id ${id}:`, error);
      // 将错误继续向上抛出
      throw error;
    }
  }
};

File: ruoyi-admin/src/main/java/com/ruoyi/tmdb_movie/service/MovieService.java
package com.ruoyi.tmdb_movie.service;

import com.alibaba.fastjson2.JSON;
import com.ruoyi.tmdb_movie.config.RestTemplateConfig;
import com.ruoyi.tmdb_movie.config.TmdbConfig;
import com.ruoyi.tmdb_movie.domain.Movie;
import com.ruoyi.tmdb_movie.mapper.MovieMapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import java.time.Duration;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;

@Service
public class MovieService {
    @Autowired
    private MovieMapper movieMapper;

    @Autowired
    private RedisTemplate<Object, Object> redisTemplate;

    @Autowired
    private RestTemplate restTemplate;

    @Autowired
    private TmdbConfig tmdbConfig;

    private static final String REDIS_KEY = "movies:all";

    public List<Movie> getAllMovies() {
        // 1. 检查Redis缓存
        List<Movie> movies = (List<Movie>) redisTemplate.opsForValue().get(REDIS_KEY);

        if (movies != null) {
            return movies;
        }

        // 2. 查询数据库
        movies = movieMapper.selectAllMovies();

        // 3. 存入Redis（缓存1小时）
        redisTemplate.opsForValue().set(REDIS_KEY, movies, 1, TimeUnit.HOURS);

        return movies;
    }

    // 清理缓存（在数据变更时调用）
    public void clearMovieCache() {
        redisTemplate.delete(REDIS_KEY);
    }

    public String getPosterUrlByMovieId(Long movieId) {
        String url = String.format("%s/movie/%d?api_key=%s",
                tmdbConfig.getBaseUrl(), movieId, tmdbConfig.getApiKey());

        ResponseEntity<Map> response = restTemplate.getForEntity(url, Map.class);
        if (response.getStatusCode().is2xxSuccessful()) {
            Map body = response.getBody();
            if (body != null && body.containsKey("poster_path")) {
                String posterPath = (String) body.get("poster_path");
                return "https://image.tmdb.org/t/p/w500" + posterPath; // TMDB 图片基准地址
            }
        }
        return null;
    }


    private final String MOVIE_CACHE_KEY = "movie:";


    public Movie getMovieById(Integer movieId) {
        String key = MOVIE_CACHE_KEY + movieId;

        // 从 Redis 中尝试获取
        String json = (String) redisTemplate.opsForValue().get(key);
        if (json != null) {
            return JSON.parseObject(json, Movie.class);
        }

        // 不存在则从数据库查询
        Movie movie = movieMapper.getMovieById(movieId);
        if (movie != null) {
            // 存入 Redis，设置过期时间，例如 1 小时
            redisTemplate.opsForValue().set(key, JSON.toJSONString(movie), Duration.ofHours(1));
        }

        return movie;
    }
}